src/main/java/org/wildfly/security/auth/realm/ldap/SimpleDirContextFactoryBuilder.java

```diff
@@ -369,72 +399,93 @@ private SocketFactory getSocketFactory() throws NamingException {
-                env.forEach((key, value) -> {
-                    if (value instanceof Object[]) {
-                        log.debugf("    Property [%s] with values %s", key, Arrays.deepToString((Object[]) value));
-                    } else {
-                        log.debugf("    Property [%s] with value [%s]", key, value);
-                    }
-                });
-            }
+                if (log.isDebugEnabled()) {
+                    log.debugf("Creating [" + InitialDirContext.class + "] with environment:");
+                    env.forEach((key, value) -> {
+                        if (value instanceof Object[]) {
+                            log.debugf("    Property [%s] with values %s", key, Arrays.deepToString((Object[]) value));
+                        } else {
+                            log.debugf("    Property [%s] with value [%s]", key, value);
+                        }
+                    });
+                }
 
-            InitialLdapContext initialContext;
+                InitialLdapContext initialContext;
 
-            try {
-                if (socketFactory != null) ThreadLocalSSLSocketFactory.set(socketFactory);
-                initialContext = new InitialLdapContext(env, null);
-            } catch (NamingException ne) {
-                log.debugf(ne, "Could not create [%s]. Failed to connect to LDAP server.", InitialLdapContext.class);
-                throw ne;
-            } finally {
-                if (socketFactory != null) ThreadLocalSSLSocketFactory.unset();
-            }
+                try {
+                    if (socketFactory != null) ThreadLocalSSLSocketFactory.set(socketFactory);
+                    initialContext = new InitialLdapContext(env, null);
+                } catch (NamingException ne) {
+                    log.debugf(ne, "Could not create [%s]. Failed to connect to LDAP server.", InitialLdapContext.class);
+                    throw ne;
+                } finally {
+                    if (socketFactory != null) ThreadLocalSSLSocketFactory.unset();
+                }
 
-            log.debugf("[%s] successfully created. Connection established to LDAP server.", initialContext);
+                log.debugf("[%s] successfully created. Connection established to LDAP server.", initialContext);
 
-            return new DelegatingLdapContext(initialContext, this::returnContext, socketFactory);
+                return new DelegatingLdapContext(initialContext, this::returnContext, socketFactory);
+            } finally{
+                setClassLoaderTo(oldClassLoader);
+            }
         }
 
         @Override
         public void returnContext(DirContext context) {
-            if (context == null) {
-                return;
-            }
+            final ClassLoader oldClassLoader = setClassLoaderTo(targetClassLoader);
+            try{
+                if (context == null) {
+                    return;
+                }
 
```

Comment:
The context==null condition could stay outside of classloader setting block. (better performance)